runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/smithy/Document.kt

```diff
@@ -4,161 +4,185 @@
-}
-
-/**
- * Class representing document `bool` type
- */
-data class SmithyBool(val value: Boolean) : Document() {
-    override fun toString(): String = when (value) {
-        true -> "true"
-        false -> "false"
-    }
-}
+import kotlinx.serialization.json.*
 
 /**
- * Class representing document `string` type
- */
-data class SmithyString(val value: String) : Document() {
-    override fun toString(): String = "\"$value\""
-}
-
-/**
- * Class representing document numeric types.
+ * A `Document` is used to store arbitrary or unstructured data.
+ *
+ * The provided casting functions (eg. [asInt], [asMap]) allow callers to unwrap contents of the Document at runtime.
  *
- * Creates a Document from a number literal: Int, Long, Short, Byte, Float, Double
+ * Document serializes to and from JSON, both over the wire and when converting to or from caller classes that implement
+ * [Transformable].
  */
-class SmithyNumber(val content: Number) : Document() {
+sealed class Document {
+    companion object {
+        /**
+         * Builds a Document using the given [Builder].
+         */
+        operator fun invoke(init: Builder.() -> Unit): Document {
+            val builder = Builder()
+            builder.init()
+            return Map(builder.content)
+        }
+
+        operator fun invoke(value: kotlin.Number): Document = Number(value)
+        operator fun invoke(value: kotlin.String): Document = String(value)
+        operator fun invoke(value: kotlin.Boolean): Document = Boolean(value)
+        operator fun invoke(value: Transformable): Document = fromString(value.serialize())
+
+        /**
+         * Creates a Document from a serialized value.
+         */
+        fun fromString(value: kotlin.String): Document =
+            fromJsonElement(
+                Json.parseToJsonElement(value)
+            )
+
+        private fun fromJsonElement(value: JsonElement): Document =
+            when (value) {
+                is JsonPrimitive -> fromJsonPrimitive(value)
+                is JsonArray -> List(value.map { fromJsonElement(it) })
+                is JsonObject -> Map(value.mapValues { fromJsonElement(it.value) })
+                JsonNull -> Null
+            }
+
+        private fun fromJsonPrimitive(value: JsonPrimitive): Document =
+            when {
+                value.intOrNull != null -> Number(value.int)
+                value.longOrNull != null -> Number(value.long)
+                value.floatOrNull != null -> Number(value.float)
+                value.doubleOrNull != null -> Number(value.double)
+                value.isString -> String(value.content)
+                value.booleanOrNull != null -> Boolean(value.boolean)
+                else -> throw IllegalArgumentException("json value $value could not be deserialized")
+            }
+
+        /**
+         * Creates a [List] from the provided values.
+         */
+        fun listOf(vararg values: Any?): Document =
```

Comment:
I don't even think we need this.

If the builder has an infix function to a list like:
```kotlin
infix fun kotlin.String.to(value: List<Any?>)
```

then the builder can work like:
```kotlin
Document {
    "myList" to listOf(12, null, true)
}
```

Reply:
I actually had that at first, but then you can't construct a Document outside of the builder directly as an array eg.

```kotlin
val Document = Document.listOf(1, null, "blah")
```